onnx/defs/schema.cc

```diff
@@ -94,6 +94,71 @@ OpSchemaRegistry* OpSchemaRegistry::Instance() {
   return &instance;
 }
 
+void OpSchema::CheckInputOutputType(struct InferenceContext& ctx) const {
+  std::unordered_map<std::string, DataType> type_constraints;
+  // check all input types
+  for (size_t in_idx = 0; in_idx < ctx.getNumInputs() && in_idx < inputs_.size(); ++in_idx) {
+    const auto& param      = inputs_[in_idx];
+    const auto& type_str   = param.GetTypeStr();
+    const auto& param_type = ctx.getInputType(in_idx);
+    const auto& all_types  = param.GetTypes();
+    if (nullptr == param_type ||
+       (param_type->has_tensor_type() && param_type->tensor_type().elem_type() == TensorProto::UNDEFINED) || 
+       (param_type->has_sequence_type() && param_type->sequence_type().elem_type().tensor_type().elem_type() == TensorProto::UNDEFINED)) {
+      // fail_check(param.GetName(), " has undefined type");
+      continue;
+    } else if (all_types.find(Utils::DataTypeUtils::ToType(*param_type)) == all_types.end()) {
+      fail_check(param.GetName(), " has unsupported type");
+    }
+    if (param.GetIsHomogeneous()) {
+      const auto& type_proto = Utils::DataTypeUtils::ToType(*param_type);
+      if (type_constraints.find(type_str) == type_constraints.end()) {
+        type_constraints[type_str] = type_proto;
+      } else if (type_constraints[type_str] != type_proto) {
+        fail_check(param.GetName(), " has inconsistent type");
+      }
+    }
+  }//for inputs
+  //check all output types
+  for (size_t out_idx = 0; out_idx < ctx.getNumOutputs() && out_idx < outputs_.size(); ++out_idx) {
+    const auto& param      = outputs_[out_idx];
+    const auto& type_str   = param.GetTypeStr();
+    const auto& param_type = ctx.getOutputType(out_idx);
+    const auto& all_types  = param.GetTypes();
+    bool output_type_found = true;
+    //infer type if necessary
```

Comment:
I think it would be better to do:
```C++
if (param_type->value_case() == VALUE_NOT_SET) {
   if (all_types.size() == 1 || type_constraints.find(type_str) != type_constraints.end()) {
      *param_type = Utils::DataTypeUtils::ToTypeProto(all_types.size() == 1 ? *all_types.begin() : type_constraints[type_str]);
   }
   continue;
}
```

Reply:
It helps handle other type-cases (ONNXML has more types than ONNX: ONNX has only tensor-type and sequence-type).

Reply:
Good suggestion, thanks!

Reply:
Done.